### DISCUSSION TITLE
CodeAnalysis MaybeNull/AllowNull on Collections

### DISCUSSION THREAD

**Discussion by meixger** [mganss/XmlSchemaClassGenerator#539] [Q&A] [answered]
Why doesn't the unit test expect MaybeNull and AllowNull on empty collections?
In the end, the property can be set to `null` at runtime.

https://github.com/mganss/XmlSchemaClassGenerator/blob/44234f600229f77dc00f7fd07669aaab4877e20a/XmlSchemaClassGenerator.Tests/xsd/nullablereferenceattributes/nullablereference.xsd#L16-L22

https://github.com/mganss/XmlSchemaClassGenerator/blob/44234f600229f77dc00f7fd07669aaab4877e20a/XmlSchemaClassGenerator.Tests/XmlTests.cs#L2691-L2691

```c#
[XmlIgnoreAttribute()]
private String[] _text;

// generated code is missing analysis attributes
// [AllowNullAttribute()]
// [MaybeNullAttribute()]
[XmlElementAttribute("Text")]
public String[] Text
{
    get
    {
        return _text;
    }
    set
    {
        _text = value;
    }
}
```

PS: The XmlSerializer is inconsitent when initializing `List<>` and `Array`: https://github.com/mganss/XmlSchemaClassGenerator/issues/538#issuecomment-2520523000

**Comment by meixger** [comment]
https://github.com/mganss/XmlSchemaClassGenerator/blob/269889db643b24765a71510ec34180d68bb870a9/XmlSchemaClassGenerator/TypeModel.cs#L603-L612

**Comment by mganss** [comment]
I think this may stem from a time prior to `collectionSettersMode`. Allowing null on collections is only necessary when the setter is not private. If it's private it can never be null.

**Comment by mganss** [comment, marked as answer]
I'm thinking to change the cited code like this:

```c#
private bool IsNullable => DefaultValue == null && !IsRequired;
private bool IsValueType => PropertyType is EnumModel || (PropertyType is SimpleModel model && model.ValueType.IsValueType);
private bool IsNullableValueType => IsNullable && !IsEnumerable && IsValueType;
private bool IsNullableReferenceType => IsNullable && (!IsEnumerable || !IsPrivateSetter) && (PropertyType is ClassModel || (PropertyType is SimpleModel model && !model.ValueType.IsValueType));
```

**Reply by meixger** [reply]
Yes, this would work for my use case

https://github.com/mganss/XmlSchemaClassGenerator/compare/master...meixger:XmlSchemaClassGenerator:nullable-reference-collections-exp

**Comment by mganss** [comment]
OK done in https://github.com/mganss/XmlSchemaClassGenerator/commit/3a978253787b2f54acf98182ae971705f0272c3b
